From: Sanchit Garg <sancgarg@linux.vnet.ibm.com>
Subject: [PATCH 093/111]     net/9p: Return error on read with NULL buffer
Git-commit: f6ac55b6c156cebf750376dc08e06ffdade82717
Patch-Mainline: 2.6.37
References: FATE#311639

    This patch ensures that a read(fd, NULL, 10) returns  EFAULT on a 9p file.

	upstream commit f6ac55b6c156cebf750376dc08e06ffdade82717

    Signed-off-by: Sanchit Garg <sancgarg@linux.vnet.ibm.com>
    Signed-off-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
    Signed-off-by: Eric Van Hensbergen <ericvh@gmail.com>
Signed-off-by: Harsh Prateek Bora <harsh@linux.vnet.ibm.com>
Acked-by: Bruce Rogers <brogers@suse.com>
---
 net/9p/client.c |    5 +----
 1 files changed, 1 insertions(+), 4 deletions(-)

diff --git a/net/9p/client.c b/net/9p/client.c
index 3742840..bee667d 100644
--- a/net/9p/client.c
+++ b/net/9p/client.c
@@ -1285,16 +1285,13 @@ p9_client_read(struct p9_fid *fid, char *data, char __user *udata, u64 offset,
 
 	if (data) {
 		memmove(data, dataptr, count);
-	}
-
-	if (udata) {
+	} else {
 		err = copy_to_user(udata, dataptr, count);
 		if (err) {
 			err = -EFAULT;
 			goto free_and_error;
 		}
 	}
-
 	p9_free_req(clnt, req);
 	return count;
 
-- 
1.7.1.1

